### PR TITLE
Mosquitto 2.x conan2

### PR DESCRIPTION
**Pull request by franramirez688**
Original PR coming from https://github.com/conan-io/conan-center-index/pull/17007

Testing some local changes to avoid polluting the original pull request.

### CODE REVIEW

File: recipes/mosquitto/2.x/conanfile.py [hunk 1]
```diff
@@ -76,61 +70,38 @@ def requirements(self):
...
-        # this would actually be the right patch for static builds also, but this would have other side effects, so
-        if(self.settings.os in ["iOS", "watchOS", "tvOS"]):
-            replace_in_file(self, os.path.join(self._source_subfolder, "lib", "CMakeLists.txt"), "SHARED", "")
-            replace_in_file(self, os.path.join(self._source_subfolder, "lib", "cpp", "CMakeLists.txt"), "SHARED", "")
+            tc.variables["WITH_CJSON"] = self.options.get_safe("cjson")
+        tc.variables["WITH_BROKER"] = self.options.broker
+        tc.variables["WITH_APPS"] = self.options.apps
+        tc.variables["WITH_PLUGINS"] = False
+        tc.variables["WITH_LIB_CPP"] = self.options.build_cpp
+        tc.variables["WITH_THREADING"] = self.settings.compiler != "msvc" and self.options.threading
```
Comment (prince-chrismc) [inline recipes/mosquitto/2.x/conanfile.py:89]: For 1.x and 2.0 support :)

```suggestion
        tc.variables["WITH_THREADING"] = is_msvc(self) and self.options.threading
```

File: recipes/mosquitto/2.x/conanfile.py [hunk 2]
```diff
@@ -139,9 +110,9 @@ def package(self):
             rm(self, "*.dylib", os.path.join(self.package_folder, "lib"))
             rm(self, "*.dll", os.path.join(self.package_folder, "bin"))
         elif self.options.shared and self.settings.compiler == "Visual Studio":
-            self.copy("mosquitto.lib", src=os.path.join(self._build_subfolder, "lib"), dst="lib")
+            copy(self, "mosquitto.lib", src=os.path.join(self.folders.build, "lib"), dst="lib")
             if self.options.build_cpp:
-                self.copy("mosquittopp.lib", src=os.path.join(self._build_subfolder, "lib"), dst="lib")
+                copy(self, "mosquittopp.lib", src=os.path.join(self.folders.build, "lib"), dst="lib")
 
     def package_info(self):
```
Comment (prince-chrismc) [inline recipes/mosquitto/2.x/conanfile.py:117]: There are some pkg_config properties that need to be updated for the new Generators :)

File: recipes/mosquitto/2.x/conanfile.py [hunk 3]
```diff
@@ -40,32 +42,25 @@ class Mosquitto(ConanFile):
...
-            del self.settings.compiler.libcxx
-            del self.settings.compiler.cppstd
+            self.options.rm_safe("compiler.libcxx")
+            self.options.rm_safe("compiler.cppstd")
+
+    def layout(self):
+        cmake_layout(self, src_folder="src")
 
     def requirements(self):
         if self.options.ssl:
```
Comment (uilianries) [inline recipes/mosquitto/2.x/conanfile.py:66]: You can change the openssl version to the range version `[>=1.1 <4]` 

For Cmake and OpenSSL we are adopting ranges: https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/dependencies.md#version-ranges

**Comment by franramirez688** [comment]
I made a new commit, but it does not appear here. GitHub is not updating the full Git history.

**Comment by franramirez688** [comment]
Sorry for forcing the update, but Github was not updating anything.

**Comment by uilianries** [comment]
@franramirez688 last error is because those libmosquitto.lib (DLL wrapper) were not installed. Or, they are in another folder and need to be copied, or need to use CMAKE_EXPORT_ALL_SYMBOLS.

File: recipes/mosquitto/2.x/conanfile.py [hunk 4]
```diff
@@ -40,113 +42,90 @@ class Mosquitto(ConanFile):
...
+        tc.variables["WITH_BROKER"] = self.options.broker
+        tc.variables["WITH_APPS"] = self.options.apps
+        tc.variables["WITH_PLUGINS"] = False
+        tc.variables["WITH_LIB_CPP"] = self.options.build_cpp
+        tc.variables["WITH_THREADING"] = not is_msvc(self) and self.options.threading
+        tc.variables["WITH_WEBSOCKETS"] = self.options.get_safe("websockets", False)
+        tc.variables["STATIC_WEBSOCKETS"] = self.options.get_safe("websockets", False) and not self.dependencies["libwebsockets"].options.shared
+        tc.variables["DOCUMENTATION"] = False
+        tc.variables["CMAKE_INSTALL_SYSCONFDIR"] = os.path.join(self.package_folder, "res").replace("\\", "/")
+        tc.variables["CMAKE_EXPORT_ALL_SYMBOLS"] = True
```
Comment (uilianries) [inline recipes/mosquitto/2.x/conanfile.py:95]:
```suggestion
        tc.variables["CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS"] = True
```

@franramirez688 my bad, the correct directive is [CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS](https://cmake.org/cmake/help/v3.4/variable/CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS.html)

**Review by uilianries — APPROVED** [review]
LGTM

File: recipes/mosquitto/2.x/test_package/test_package.c
Comment (RubenRBS) [inline recipes/mosquitto/2.x/test_package/test_package.c:1]: I would simplify this to its barest minimum enough to check that the package was created correctly

**Review by memsharded — APPROVED** [review]

File: recipes/mosquitto/2.x/test_package/test_package.c
Comment (danicriss) [reply recipes/mosquitto/2.x/test_package/test_package.c:1]: It was barest minimum and I've modified it in https://github.com/conan-io/conan-center-index/pull/17507/commits/29d0acd54e31b2bf888fecae4c6e3f0ae8e942ad. For background, mosquitto v1.6.12 had already been ported to conan2 in https://github.com/conan-io/conan-center-index/pull/15714. I've just copied its test for mosquitto versions 2.x (v2.0.3 - 2.0.15 as of now)

The rationale is that we've had the bare minimum for an internal package that we built in my company, and during development we missed linking some of its libraries iirc. But we had caught that way too late (at usage point) so we found value in adding more substance to the testing of the conan package build

**Pull request merged**